Issue/843 add autocomplete suggestions

theck13 opened the pull request:
### Fix
Add autocomplete suggestion to the search input field to close #843.  See the screenshots and animation below for illustration.

![843_add_autocomplete_suggestions](https://user-images.githubusercontent.com/3827611/67817934-11e74e80-fa75-11e9-8858-ccf1fbd6b5e5.png)

<a href="https://user-images.githubusercontent.com/3827611/67818109-a2be2a00-fa75-11e9-9e4e-40cdc7a0f2c5.gif"><img src="https://user-images.githubusercontent.com/3827611/67818109-a2be2a00-fa75-11e9-9e4e-40cdc7a0f2c5.gif" width="320"></a>

### Test
1. Tap ***Search*** action in top app bar.
2. Notice note list is blank.
3. Enter text into search field.
4. Notice suggestions appear based on entered text.
5. Tap any suggestion.
6. Notice suggestion text is inserted into search field.
7. Notice note list is shown based on inserted text.

### Review
Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

AmandaRiu commented on Simplenote/src/main/java/com/automattic/simplenote/NoteListFragment.java:
```diff
@@ -673,19 +691,31 @@ public void setNoteSelected(String selectedNoteID) {
+        }
+
+        if (searchString.isEmpty()) {
+            // TODO: Get history items.
```
Just verifying this `// TODO` isn't an oversight

AmandaRiu commented:
Ready for merge, just waiting on design review.

SylvesterWilmott commented:
Looks great @theck13 

Can you explain how the tag suggestion order works? I created tags named "sa" "sb" "sc" .. etc and when searching "s" one of my older tags "plants" is top of my list, is this because it's used in more notes?

Also when I type "tag:" or "tag:[x]" in the search bar, I no longer have suggestions.

theck13 commented:
> Can you explain how the tag suggestion order works?

The tags are ordered by their key, which is effectively the created date, in descending order.

> Also when I type "tag:" or "tag:[x]" in the search bar, I no longer have suggestions.

Unless you have note or tag with "tag:" in it, searching for "tag:" will show no results.

> In landscape orientation, I also noticed the "X" button in the search bar is not aligned to the right, is this an artefact from our tablet view?

The search view is part of the top app bar and not either of the note list or note editor/preview panes below it.  The "X" button is part of that view and isn't modified by these changes.

> Can you also add some tablet screens to the PR too?

Before|After
-|-
![843_add_autocomplete_suggestions_tablet_before_01](https://user-images.githubusercontent.com/3827611/67961986-153c2080-fbc2-11e9-8cbe-4f2e348716a2.png)|![843_add_autocomplete_suggestions_tablet_after_01](https://user-images.githubusercontent.com/3827611/67961990-18371100-fbc2-11e9-9b0e-fffb0480abb1.png)
![843_add_autocomplete_suggestions_tablet_before_02](https://user-images.githubusercontent.com/3827611/67962004-1e2cf200-fbc2-11e9-9bd6-2c3e7e73fac9.png)|![843_add_autocomplete_suggestions_tablet_after_02](https://user-images.githubusercontent.com/3827611/67962012-2127e280-fbc2-11e9-884b-8e08803db199.png)
![843_add_autocomplete_suggestions_tablet_before_03](https://user-images.githubusercontent.com/3827611/67962017-2422d300-fbc2-11e9-8c64-4441b92b76c4.png)|![843_add_autocomplete_suggestions_tablet_after_03](https://user-images.githubusercontent.com/3827611/67962028-26852d00-fbc2-11e9-852c-de72226c8f8b.png)

theck13 commented:
@SylvesterWilmott and I discussed how to handle the tag suggestion sorting and behavior.  We came up with this:

<blockquote>
<ul>
<li>
Tag suggestions are always ordered by most used to least used (i.e. tags assigned to the most notes are listed first/top).
</li>
<li>
All tags are shown when `tag:` is entered.
</li>
</blockquote>

theck13 commented:
<blockquote>
<ul>
<li>
Tag suggestions are always ordered by most used to least used (i.e. tags assigned to the most notes are listed first/top).
</li>
</blockquote>

Updated tag suggestions order in https://github.com/Automattic/simplenote-android/pull/865/commits/78a16141d83709d0cec4f82f6705da3b65f40094.

<blockquote>
<li>
All tags are shown when <code>tag:</code> is entered.
</li>
</blockquote>

Updated showing all tags with `tag:` in https://github.com/Automattic/simplenote-android/pull/865/commits/205711fcbd0ece9cab61e2dd5727fdea1503c996.

theck13 commented:
I updated the behavior when tapping a tag from the suggestions in a40224d1 so that multiple tags can be searched at once.  See the animation below for illustration.

<a href="https://user-images.githubusercontent.com/3827611/68172654-93d3ed80-ff35-11e9-8651-cb03aed6dfe7.gif"><img src="https://user-images.githubusercontent.com/3827611/68172654-93d3ed80-ff35-11e9-8651-cb03aed6dfe7.gif" width="320"></a>

SylvesterWilmott commented:
Looks good @theck13
